File: final/internal/infra/infra.go
// Package infra 管理所有外部基础设施连接：Milvus / PostgreSQL / Elasticsearch / Kafka
// 每个连接失败时优雅降级，不影响应用启动。
package infra

import (
	"context"
	"database/sql"
	"encoding/json"
	"final/config"
	"fmt"
	"log"
	"strings"
	"time"

	es "github.com/elastic/go-elasticsearch/v8"
	_ "github.com/lib/pq"
	milvusClient "github.com/milvus-io/milvus-sdk-go/v2/client"
	"github.com/milvus-io/milvus-sdk-go/v2/entity"
	"github.com/segmentio/kafka-go"
)

// Status 记录各基础设施的连接状态
type Status struct {
	Milvus string `json:"milvus"`
	PG     string `json:"pg"`
	ES     string `json:"elasticsearch"`
	Kafka  string `json:"kafka"`
}

// Infrastructure 持有所有外部连接句柄
type Infrastructure struct {
	cfg    *config.APIConfig
	milvus milvusClient.Client
	pg     *sql.DB
	es     *es.Client
	kafkaW *kafka.Writer
	Ready  Status
}

// New 尝试连接所有基础设施，失败则降级为内存模式。
func New(cfg *config.APIConfig) *Infrastructure {
	inf := &Infrastructure{cfg: cfg}
	inf.connectMilvus()
	inf.connectPostgres()
	inf.connectES()
	inf.connectKafka()
	return inf
}

// ─────────────────────────────── 连接初始化 ───────────────────────────────

func (inf *Infrastructure) connectMilvus() {
	mc, err := milvusClient.NewClient(context.Background(), milvusClient.Config{
		Address: inf.cfg.MilvusAddr(),
	})
	if err != nil {
		log.Printf("⚠️  Milvus 连接失败: %v (将使用内存向量库)", err)
		inf.Ready.Milvus = "disconnected"
		return
	}
	inf.milvus = mc
	inf.Ready.Milvus = "connected"
	log.Println("✅ Milvus 已连接:", inf.cfg.MilvusAddr())
}

func (inf *Infrastructure) connectPostgres() {
	pg, err := sql.Open("postgres", inf.cfg.PGDSN())
	if err != nil {
		log.Printf("⚠️  PostgreSQL 打开失败: %v", err)
		inf.Ready.PG = "disconnected"
		return
	}
	if err := pg.Ping(); err != nil {
		log.Printf("⚠️  PostgreSQL Ping 失败: %v", err)
		inf.Ready.PG = "disconnected"
		return
	}
	inf.pg = pg
	inf.Ready.PG = "connected"
	inf.initPGSchema()
	log.Println("✅ PostgreSQL 已连接:", inf.cfg.PGDSN())
}

func (inf *Infrastructure) connectES() {
	esCfg := es.Config{
		Addresses: inf.cfg.ESAddresses,
		Username:  inf.cfg.ESUsername,
		Password:  inf.cfg.ESPassword,
	}
	esClient, err := es.NewClient(esCfg)
	if err != nil {
		log.Printf("⚠️  Elasticsearch 连接失败: %v", err)
		inf.Ready.ES = "disconnected"
		return
	}
	res, err := esClient.Info()
	if err != nil {
		log.Printf("⚠️  Elasticsearch Ping 失败: %v", err)
		inf.Ready.ES = "disconnected"
		return
	}
	res.Body.Close()
	inf.es = esClient
	inf.Ready.ES = "connected"
	log.Println("✅ Elasticsearch 已连接:", inf.cfg.ESAddresses)
}

func (inf *Infrastructure) connectKafka() {
	inf.kafkaW = &kafka.Writer{
		Addr:         kafka.TCP(inf.cfg.KafkaBrokers...),
		Topic:        inf.cfg.KafkaTopic,
		Balancer:     &kafka.LeastBytes{},
		BatchTimeout: 10 * time.Millisecond,
	}
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	conn, err := kafka.DialLeader(ctx, "tcp", inf.cfg.KafkaBrokers[0], inf.cfg.KafkaTopic, 0)
	if err != nil {
		log.Printf("⚠️  Kafka 连接失败: %v (事件将输出到日志)", err)
		inf.Ready.Kafka = "disconnected"
		return
	}
	conn.Close()
	inf.Ready.Kafka = "connected"
	log.Println("✅ Kafka 已连接:", inf.cfg.KafkaBrokers)
}

// ─────────────────────────────── PostgreSQL ───────────────────────────────

// initPGSchema 建表（幂等）
func (inf *Infrastructure) initPGSchema() {
	if inf.pg == nil {
		return
	}
	ddls := []string{
		`CREATE TABLE IF NOT EXISTS user_preferences (
			user_id    TEXT NOT NULL,
			key        TEXT NOT NULL,
			value      TEXT NOT NULL,
			updated_at TIMESTAMP DEFAULT NOW(),
			PRIMARY KEY (user_id, key)
		)`,
		`CREATE TABLE IF NOT EXISTS task_snapshots (
			task_id    TEXT PRIMARY KEY,
			state      JSONB NOT NULL,
			created_at TIMESTAMP DEFAULT NOW()
		)`,
		`CREATE TABLE IF NOT EXISTS chat_history (
			id         SERIAL PRIMARY KEY,
			role       TEXT NOT NULL,
			content    TEXT NOT NULL,
			created_at TIMESTAMP DEFAULT NOW()
		)`,
		`CREATE TABLE IF NOT EXISTS long_term_memory (
			id          SERIAL PRIMARY KEY,
			content     TEXT NOT NULL,
			importance  FLOAT NOT NULL DEFAULT 0.5,
			embedding   JSONB,
			created_at  TIMESTAMP DEFAULT NOW()
		)`,
	}
	for _, ddl := range ddls {
		if _, err := inf.pg.Exec(ddl); err != nil {
			log.Printf("⚠️  PG 建表失败: %v", err)
		}
	}
	log.Println("✅ PostgreSQL 表结构已初始化")
}

// SavePreference 持久化用户偏好到 PostgreSQL（upsert）
func (inf *Infrastructure) SavePreference(userID, key, value string) {
	if inf.pg == nil {
		return
	}
	_, err := inf.pg.Exec(
		`INSERT INTO user_preferences (user_id, key, value) VALUES ($1, $2, $3)
		 ON CONFLICT (user_id, key) DO UPDATE SET value = $3, updated_at = NOW()`,
		userID, key, value,
	)
	if err != nil {
		log.Printf("⚠️  偏好保存到 PG 失败: %v", err)
	}
}

// SaveSnapshot 持久化任务快照到 PostgreSQL（upsert）
func (inf *Infrastructure) SaveSnapshot(taskID string, stateJSON []byte) {
	if inf.pg == nil {
		return
	}
	_, err := inf.pg.Exec(
		`INSERT INTO task_snapshots (task_id, state) VALUES ($1, $2)
		 ON CONFLICT (task_id) DO UPDATE SET state = $2, created_at = NOW()`,
		taskID, stateJSON,
	)
	if err != nil {
		log.Printf("⚠️  快照保存到 PG 失败: %v", err)
	}
}

// LoadPreferences 从 PostgreSQL 加载指定用户的全部偏好，返回 map[key]value
func (inf *Infrastructure) LoadPreferences(userID string) map[string]string {
	result := make(map[string]string)
	if inf.pg == nil {
		return result
	}
	rows, err := inf.pg.Query(`SELECT key, value FROM user_preferences WHERE user_id = $1`, userID)
	if err != nil {
		log.Printf("⚠️  加载偏好失败: %v", err)
		return result
	}
	defer rows.Close()
	for rows.Next() {
		var k, v string
		if err := rows.Scan(&k, &v); err == nil {
			result[k] = v
		}
	}
	return result
}

// LongTermRow 是从 PG 读取的长期记忆行
type LongTermRow struct {
	ID         int
	Content    string
	Importance float64
	Embedding  []float64
}

// SaveLongTermItem 将一条长期记忆持久化到 PostgreSQL，返回数据库自增 ID
func (inf *Infrastructure) SaveLongTermItem(content string, importance float64, embeddingJSON []byte) int {
	if inf.pg == nil {
		return -1
	}
	var id int
	err := inf.pg.QueryRow(
		`INSERT INTO long_term_memory (content, importance, embedding) VALUES ($1, $2, $3) RETURNING id`,
		content, importance, embeddingJSON,
	).Scan(&id)
	if err != nil {
		log.Printf("⚠️  长期记忆保存失败: %v", err)
		return -1
	}
	return id
}

// LoadLongTermItems 从 PostgreSQL 加载全部长期记忆条目
func (inf *Infrastructure) LoadLongTermItems() []LongTermRow {
	if inf.pg == nil {
		return nil
	}
	rows, err := inf.pg.Query(`SELECT id, content, importance, embedding FROM long_term_memory ORDER BY id`)
	if err != nil {
		log.Printf("⚠️  加载长期记忆失败: %v", err)
		return nil
	}
	defer rows.Close()
	var items []LongTermRow
	for rows.Next() {
		var row LongTermRow
		var embJSON []byte
		if err := rows.Scan(&row.ID, &row.Content, &row.Importance, &embJSON); err != nil {
			continue
		}
		if len(embJSON) > 0 {
			json.Unmarshal(embJSON, &row.Embedding)
		}
		items = append(items, row)
	}
	return items
}

// ─────────────────────────────── Elasticsearch ───────────────────────────

// SearchES 在 Elasticsearch 中执行 JSON 查询，返回原始响应字符串
func (inf *Infrastructure) SearchES(index, queryJSON string) (string, error) {
	if inf.es == nil {
		return "", fmt.Errorf("elasticsearch not connected")
	}
	resp, err := inf.es.Search(
		inf.es.Search.WithIndex(index),
		inf.es.Search.WithBody(strings.NewReader(queryJSON)),
	)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	var result map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&result)
	data, _ := json.Marshal(result)
	return string(data), nil
}

// ─────────────────────────────── Milvus ──────────────────────────────────

// MilvusSearch 在 Milvus 中进行向量近邻搜索，返回匹配文档 ID 列表
func (inf *Infrastructure) MilvusSearch(collection string, vector []float32, topK int) ([]int64, error) {
	if inf.milvus == nil {
		return nil, fmt.Errorf("milvus not connected")
	}
	sp, _ := entity.NewIndexFlatSearchParam()
	results, err := inf.milvus.Search(
		context.Background(), collection, []string{},
		"", []string{"content"},
		[]entity.Vector{entity.FloatVector(vector)},
		"embedding", entity.L2,
		topK, sp,
	)
	if err != nil {
		return nil, err
	}
	var ids []int64
	for _, r := range results {
		for _, id := range r.IDs.FieldData().GetScalars().GetLongData().Data {
			ids = append(ids, id)
		}
	}
	return ids, nil
}

// ─────────────────────────────── Kafka ───────────────────────────────────

// PublishEvent 向 Kafka 发布事件；未连接时退化为日志输出
func (inf *Infrastructure) PublishEvent(eventType, payload string) {
	msg := kafka.Message{
		Key:   []byte(eventType),
		Value: []byte(payload),
	}
	if inf.Ready.Kafka == "connected" {
		if err := inf.kafkaW.WriteMessages(context.Background(), msg); err != nil {
			log.Printf("⚠️  Kafka 写入失败: %v", err)
		}
	} else {
		log.Printf("📋 [Kafka-fallback] %s: %s", eventType, payload)
	}
}

// ─────────────────────────────── 生命周期 ────────────────────────────────

// Close 释放所有连接资源
func (inf *Infrastructure) Close() {
	if inf.milvus != nil {
		inf.milvus.Close()
	}
	if inf.pg != nil {
		inf.pg.Close()
	}
	if inf.kafkaW != nil {
		inf.kafkaW.Close()
	}
}

File: final/internal/infra/infra.py
# infra — 管理所有外部基础设施连接：Milvus / PostgreSQL / Elasticsearch / Kafka
# 每个连接失败时优雅降级，不影响应用启动。
#
# 业务持久化逻辑统一收敛到 internal.repo 包：Infrastructure 仅负责连接生命周期
# (connect / schema bootstrap / health) 与跨域装配 self.repo 仓储入口。
import json
import logging
from dataclasses import dataclass
from types import SimpleNamespace
from typing import Any, Iterable, List, Optional, Sequence, Tuple

from config.config import APIConfig
from internal.repo import chathistory, eventbus, longterm, preference, ragchunk, snapshot

logger = logging.getLogger(__name__)

# 尝试导入可选依赖，失败则标记不可用
try:
    import psycopg2
    _HAS_PG = True
except ImportError:
    _HAS_PG = False

try:
    from elasticsearch import Elasticsearch
    _HAS_ES = True
except ImportError:
    _HAS_ES = False

try:
    from kafka import KafkaProducer
    _HAS_KAFKA = True
except ImportError:
    _HAS_KAFKA = False

try:
    from pymilvus import DataType, MilvusClient
    _HAS_MILVUS = True
except ImportError:
    DataType = None  # type: ignore
    _HAS_MILVUS = False


# 默认 RAG 集合名（与 main 分支 Go 实现 internal/infrastructure/persistence/ragchunk 对齐）
RAG_COLLECTION = "rag_chunks"
# 索引参数（与 Go 端 entity.NewIndexIvfFlat(entity.L2, 128) 对齐）
_RAG_INDEX_TYPE = "IVF_FLAT"
_RAG_METRIC_TYPE = "L2"
_RAG_INDEX_NLIST = 128
# content 字段最大长度（与 Go 端 max_length=4096 对齐）
_RAG_CONTENT_MAX_LEN = 4096


@dataclass
class Status:
    milvus: str = "disconnected"
    postgresql: str = "disconnected"
    elasticsearch: str = "disconnected"
    kafka: str = "disconnected"


@dataclass
class LongTermRow:
    id: int
    content: str
    importance: float
    embedding: Optional[List[float]] = None


# ─────────────────────── repo 适配层 ──────────────────────────────────────
# repo 期望 internal.platform 风格的 client（含 is_real()/query/exec/conn 等）。
# 这里把 Infrastructure 已经持有的 raw 连接句柄包成符合接口的 thin adapter，
# 避免重复 connect。

class _PGAdapter:
    """把 psycopg2 raw connection 包成 PostgresClient-like 接口供 repo 使用。"""

    def __init__(self, conn):
        self._conn = conn

    def is_real(self) -> bool:
        return self._conn is not None

    @property
    def conn(self):
        return self._conn

    def query(self, sql: str, params: Optional[Sequence[Any]] = None) -> List[Tuple[Any, ...]]:
        if self._conn is None:
            return []
        try:
            with self._conn.cursor() as cur:
                cur.execute(sql, params or ())
                return list(cur.fetchall())
        except Exception as e:
            logger.warning("⚠️  PG query 失败: %s", e)
            return []

    def query_one(self, sql: str, params: Optional[Sequence[Any]] = None) -> Optional[Tuple[Any, ...]]:
        if self._conn is None:
            return None
        try:
            with self._conn.cursor() as cur:
                cur.execute(sql, params or ())
                return cur.fetchone()
        except Exception as e:
            logger.warning("⚠️  PG query_one 失败: %s", e)
            return None

    def exec(self, sql: str, params: Optional[Sequence[Any]] = None) -> int:
        if self._conn is None:
            return -1
        try:
            with self._conn.cursor() as cur:
                cur.execute(sql, params or ())
                return cur.rowcount
        except Exception as e:
            logger.warning("⚠️  PG exec 失败: %s", e)
            return -1

    def exec_many(self, sql: str, seq_of_params: Iterable[Sequence[Any]]) -> int:
        if self._conn is None:
            return -1
        try:
            with self._conn.cursor() as cur:
                cur.executemany(sql, list(seq_of_params))
                return cur.rowcount
        except Exception as e:
            logger.warning("⚠️  PG exec_many 失败: %s", e)
            return -1


class _ESAdapter:
    """把 raw Elasticsearch 客户端包成 ESClient-like 接口供 repo 使用。"""

    def __init__(self, es):
        self._es = es

    def is_real(self) -> bool:
        return self._es is not None

    @property
    def client(self):
        return self._es

    def index(self, index: str, doc_id: Any, body: dict) -> bool:
        if self._es is None:
            return False
        try:
            self._es.index(index=index, id=doc_id, body=body)
            return True
        except Exception as e:
            logger.warning("⚠️  ES 索引失败: %s", e)
            return False

    def search(self, index: str, body: dict) -> dict:
        if self._es is None:
            return {}
        try:
            resp = self._es.search(index=index, body=body)
            return dict(resp)
        except Exception as e:
            logger.warning("⚠️  ES 检索失败: %s", e)
            return {}

    def delete_many(self, index: str, doc_ids: List[Any]) -> None:
        if self._es is None:
            return
        for doc_id in doc_ids:
            try:
                self._es.delete(index=index, id=doc_id)
            except Exception as e:
                logger.warning("⚠️  ES 删除失败 (id=%s): %s", doc_id, e)


class _MilvusAdapter:
    """把 raw MilvusClient 包成 MilvusClientWrapper-like 接口供 repo 使用。"""

    def __init__(self, client):
        self._client = client

    def is_real(self) -> bool:
        return self._client is not None

    @property
    def client(self):
        return self._client

    def ensure_collection(self, collection_name: str, dimension: int,
                          auto_id: bool = True, enable_dynamic_field: bool = True) -> bool:
        # rag_chunks 集合由 _init_milvus_collections 在启动期建好；这里仅做存在性确认。
        if self._client is None:
            return False
        try:
            return bool(self._client.has_collection(collection_name))
        except Exception:
            return False

    def insert(self, collection_name: str, data: List[dict]) -> bool:
        if self._client is None or not data:
            return False
        try:
            self._client.insert(collection_name=collection_name, data=data)
            return True
        except Exception as e:
            logger.warning("⚠️  Milvus 插入失败: %s", e)
            return False

    def search(self, collection_name: str, query_emb: List[float], top_k: int,
               output_fields: Optional[List[str]] = None) -> List[dict]:
        if self._client is None:
            return []
        try:
            results = self._client.search(
                collection_name=collection_name,
                data=[query_emb],
                limit=top_k,
                output_fields=output_fields or ["pg_id", "content"],
            )
            hits: List[dict] = []
            if not results:
                return hits
            for result in results[0]:
                entity = result.get("entity", {}) if isinstance(result, dict) else {}
                hits.append({
                    "pg_id": entity.get("pg_id"),
                    "content": entity.get("content"),
                    "score": result.get("distance", 0.0) if isinstance(result, dict) else 0.0,
                })
            return hits
        except Exception as e:
            logger.warning("⚠️  Milvus 检索失败: %s", e)
            return []

    def delete(self, collection_name: str, filter_expr: str) -> bool:
        if self._client is None:
            return False
        try:
            self._client.delete(collection_name=collection_name, filter=filter_expr)
            return True
        except Exception as e:
            logger.warning("⚠️  Milvus 删除失败: %s", e)
            return False


class _KafkaAdapter:
    """把 raw KafkaProducer 包成 KafkaClient-like 接口供 repo.eventbus 使用。"""

    def __init__(self, producer, cfg: APIConfig, ready: Status):
        self._producer = producer
        self._cfg = cfg
        self._ready = ready

    def is_real(self) -> bool:
        return self._producer is not None and self._ready.kafka == "connected"

    def produce(self, event_type: str, payload, topic: Optional[str] = None) -> bool:
        target_topic = topic or self._cfg.kafka_topic
        if not self.is_real():
            return False
        try:
            self._producer.send(
                target_topic,
                key=event_type.encode("utf-8"),
                value=payload,
            )
            return True
        except Exception as e:
            logger.warning("⚠️  Kafka 写入失败: %s", e)
            return False


class Infrastructure:
    """持有所有外部连接句柄"""

    def __init__(self, cfg: APIConfig):
        self.cfg = cfg
        self.ready = Status()
        self._pg = None
        self._es = None
        self._kafka_producer = None
        self._milvus = None

        self._connect_postgres()
        self._connect_es()
        self._connect_kafka()
        self._connect_milvus()

        # ─── repo 装配（统一持久化入口） ────────────────────────────────
        # 业务侧应通过 inf.repo.<domain>.<method>(...) 访问；此处不重复 connect，
        # 只把已有句柄包成符合 internal.platform 接口的 thin adapter 注入 repo。
        pg_client = _PGAdapter(self._pg)
        es_client = _ESAdapter(self._es)
        milvus_client = _MilvusAdapter(self._milvus)
        kafka_client = _KafkaAdapter(self._kafka_producer, self.cfg, self.ready)

        self.repo = SimpleNamespace(
            ragchunk=ragchunk.Store(pg_client, milvus_client, es_client),
            ltm=longterm.PGRepo(pg_client),
            chat_history=chathistory.PGRepo(pg_client),
            preference=preference.PGRepo(pg_client),
            snapshot=snapshot.PGRepo(pg_client),
            events=eventbus.KafkaPublisher(kafka_client),
        )

    # ─────────────────────────────── PostgreSQL ───────────────────────────────

    def _connect_postgres(self):
        if not _HAS_PG:
            logger.warning("⚠️  psycopg2 未安装，PostgreSQL 不可用")
            return
        if not self.cfg.pg_host:
            logger.warning("⚠️  PostgreSQL 未配置")
            return
        try:
            self._pg = psycopg2.connect(self.cfg.pg_dsn())
            self._pg.autocommit = True
            with self._pg.cursor() as cur:
                cur.execute("SELECT 1")
            self.ready.postgresql = "connected"
            self._init_pg_schema()
            logger.info("✅ PostgreSQL 已连接: %s", self.cfg.pg_dsn())
        except Exception as e:
            logger.warning("⚠️  PostgreSQL 连接失败: %s", e)
            self._pg = None

    def _init_pg_schema(self):
        if not self._pg:
            return
        ddls = [
            """CREATE TABLE IF NOT EXISTS user_preferences (
                user_id    TEXT NOT NULL,
                key        TEXT NOT NULL,
                value      TEXT NOT NULL,
                updated_at TIMESTAMP DEFAULT NOW(),
                PRIMARY KEY (user_id, key)
            )""",
            """CREATE TABLE IF NOT EXISTS task_snapshots (
                task_id    TEXT PRIMARY KEY,
                state      JSONB NOT NULL,
                created_at TIMESTAMP DEFAULT NOW()
            )""",
            """CREATE TABLE IF NOT EXISTS chat_history (
                id         SERIAL PRIMARY KEY,
                role       TEXT NOT NULL,
                content    TEXT NOT NULL,
                created_at TIMESTAMP DEFAULT NOW()
            )""",
            """CREATE TABLE IF NOT EXISTS long_term_memory (
                id            SERIAL PRIMARY KEY,
                content       TEXT NOT NULL,
                importance    FLOAT NOT NULL DEFAULT 0.5,
                embedding     JSONB,
                created_at    DOUBLE PRECISION DEFAULT EXTRACT(EPOCH FROM NOW()),
                last_accessed DOUBLE PRECISION DEFAULT EXTRACT(EPOCH FROM NOW()),
                category      VARCHAR(64) NOT NULL DEFAULT '',
                tags          JSONB NOT NULL DEFAULT '[]'::jsonb,
                slot_hint     VARCHAR(64) NOT NULL DEFAULT '',
                score         DOUBLE PRECISION NOT NULL DEFAULT 0.0
            )""",
            # 老库平滑加列：每条 ALTER 独立执行，互不阻塞
            "ALTER TABLE long_term_memory ADD COLUMN IF NOT EXISTS created_at    DOUBLE PRECISION DEFAULT EXTRACT(EPOCH FROM NOW())",
            "ALTER TABLE long_term_memory ADD COLUMN IF NOT EXISTS last_accessed DOUBLE PRECISION DEFAULT EXTRACT(EPOCH FROM NOW())",
            "ALTER TABLE long_term_memory ADD COLUMN IF NOT EXISTS category      VARCHAR(64) NOT NULL DEFAULT ''",
            "ALTER TABLE long_term_memory ADD COLUMN IF NOT EXISTS tags          JSONB NOT NULL DEFAULT '[]'::jsonb",
            "ALTER TABLE long_term_memory ADD COLUMN IF NOT EXISTS slot_hint     VARCHAR(64) NOT NULL DEFAULT ''",
            "ALTER TABLE long_term_memory ADD COLUMN IF NOT EXISTS score         DOUBLE PRECISION NOT NULL DEFAULT 0.0",
            "CREATE INDEX IF NOT EXISTS idx_lti_category ON long_term_memory(category)",
            "CREATE INDEX IF NOT EXISTS idx_lti_tags     ON long_term_memory USING GIN(tags)",
            """CREATE TABLE IF NOT EXISTS rag_chunks (
                id          BIGSERIAL PRIMARY KEY,
                doc_hash    TEXT NOT NULL,
                chunk_idx   INT NOT NULL,
                content     TEXT NOT NULL,
                parent_content TEXT,
                embedding   JSONB,
                created_at  TIMESTAMP DEFAULT NOW(),
                UNIQUE (doc_hash, chunk_idx)
            )""",
            """ALTER TABLE rag_chunks ADD COLUMN IF NOT EXISTS parent_content TEXT""",
            """CREATE UNIQUE INDEX IF NOT EXISTS rag_chunks_doc_hash_chunk_idx_key
                   ON rag_chunks (doc_hash, chunk_idx)""",
        ]
        with self._pg.cursor() as cur:
            for ddl in ddls:
                try:
                    cur.execute(ddl)
                except Exception as e:
                    logger.warning("⚠️  PG 建表失败: %s", e)
        logger.info("✅ PostgreSQL 表结构已初始化")

    # ─────────────────────────────── Elasticsearch ───────────────────────────

    def _connect_es(self):
        if not _HAS_ES:
            logger.warning("⚠️  elasticsearch-py 未安装，ES 不可用")
            return
        if not self.cfg.es_addresses:
            logger.warning("⚠️  Elasticsearch 未配置")
            return
        try:
            auth = (self.cfg.es_username, self.cfg.es_password) if self.cfg.es_username else None
            self._es = Elasticsearch(
                self.cfg.es_addresses,
                basic_auth=auth,
            )
            if self._es.ping():
                self.ready.elasticsearch = "connected"
                logger.info("✅ Elasticsearch 已连接: %s", self.cfg.es_addresses)
            else:
                self._es = None
        except Exception as e:
            logger.warning("⚠️  Elasticsearch 连接失败: %s", e)
            self._es = None

    # ─────────────────────────────── Kafka ───────────────────────────────────

    def _connect_kafka(self):
        if not _HAS_KAFKA:
            logger.warning("⚠️  kafka-python 未安装，Kafka 不可用")
            return
        if not self.cfg.kafka_brokers:
            logger.warning("⚠️  Kafka 未配置")
            return
        try:
            self._kafka_producer = KafkaProducer(
                bootstrap_servers=self.cfg.kafka_brokers,
                value_serializer=lambda v: json.dumps(v).encode("utf-8"),
            )
            self.ready.kafka = "connected"
            logger.info("✅ Kafka 已连接: %s", self.cfg.kafka_brokers)
        except Exception as e:
            logger.warning("⚠️  Kafka 连接失败: %s (事件将输出到日志)", e)
            self._kafka_producer = None

    # ─────────────────────────────── Milvus ───────────────────────────────────

    def _connect_milvus(self):
        if not _HAS_MILVUS:
            logger.warning("⚠️  pymilvus 未安装，Milvus 不可用")
            self.ready.milvus = "memory-mode"
            return
        if not self.cfg.milvus_host or not self.cfg.milvus_port:
            logger.warning("⚠️  Milvus 未配置")
            self.ready.milvus = "memory-mode"
            return
        try:
            uri = f"http://{self.cfg.milvus_host}:{self.cfg.milvus_port}"
            self._milvus = MilvusClient(uri=uri)
            self.ready.milvus = "connected"
            logger.info("✅ Milvus 已连接: %s", uri)
            self._init_milvus_collections()
        except Exception as e:
            logger.warning("⚠️  Milvus 连接失败: %s (降级到内存模式)", e)
            self._milvus = None
            self.ready.milvus = "memory-mode"

    def _init_milvus_collections(self):
        if not self._milvus:
            return
        dim = int(self.cfg.rag_milvus_dim or 1024)
        try:
            if self._milvus.has_collection(RAG_COLLECTION):
                # 集合已存在：仅校验 schema，不一致只 warning，不删用户数据
                self._verify_milvus_rag_schema(RAG_COLLECTION, dim)
                return
            self._create_milvus_rag_collection(RAG_COLLECTION, dim)
        except Exception as e:
            logger.warning("⚠️  Milvus 创建集合失败: %s", e)

    def _create_milvus_rag_collection(self, collection_name: str, dim: int):
        """以显式 schema 创建 RAG 集合：pg_id (PK Int64) / content (VarChar 4096) / embedding。

        与 main 分支 Go 实现 EnsureMilvusCollection 对齐。
        """
        if not self._milvus or DataType is None:
            return
        schema = self._milvus.create_schema(auto_id=False, enable_dynamic_field=False)
        schema.add_field(field_name="pg_id", datatype=DataType.INT64, is_primary=True)
        schema.add_field(
            field_name="content",
            datatype=DataType.VARCHAR,
            max_length=_RAG_CONTENT_MAX_LEN,
        )
        schema.add_field(
            field_name="embedding", datatype=DataType.FLOAT_VECTOR, dim=dim
        )

        index_params = self._milvus.prepare_index_params()
        index_params.add_index(
            field_name="embedding",
            index_type=_RAG_INDEX_TYPE,
            metric_type=_RAG_METRIC_TYPE,
            params={"nlist": _RAG_INDEX_NLIST},
        )

        self._milvus.create_collection(
            collection_name=collection_name,
            schema=schema,
            index_params=index_params,
        )
        logger.info(
            "✅ Milvus 集合 %s 已创建 (dim=%d, index=%s, metric=%s, nlist=%d)",
            collection_name, dim, _RAG_INDEX_TYPE, _RAG_METRIC_TYPE, _RAG_INDEX_NLIST,
        )

    def _verify_milvus_rag_schema(self, collection_name: str, expected_dim: int):
        """校验已存在集合的 PK 字段名 / 向量维度，不一致只打 warning，不 drop。"""
        if not self._milvus:
            return
        try:
            desc = self._milvus.describe_collection(collection_name)
        except Exception as e:
            logger.warning("⚠️  Milvus describe_collection(%s) 失败: %s", collection_name, e)
            return

        if isinstance(desc, dict):
            fields = desc.get("fields") or []
        else:
            fields = getattr(desc, "fields", []) or []

        pk_name: Optional[str] = None
        embedding_dim: Optional[int] = None
        for f in fields:
            name = f.get("name") if isinstance(f, dict) else getattr(f, "name", None)
            is_primary = (
                f.get("is_primary") if isinstance(f, dict) else getattr(f, "is_primary", False)
            )
            if is_primary:
                pk_name = name
            if name == "embedding":
                params = (
                    f.get("params") if isinstance(f, dict) else getattr(f, "params", {})
                ) or {}
                if isinstance(params, dict):
                    raw_dim = params.get("dim")
                    try:
                        embedding_dim = int(raw_dim) if raw_dim is not None else None
                    except (TypeError, ValueError):
                        embedding_dim = None

        if pk_name is not None and pk_name != "pg_id":
            logger.warning(
                "⚠️  Milvus 集合 %s 主键字段名不一致 (expected=pg_id, actual=%s)，"
                "请手动 drop 旧集合后重新 ingest 全量数据",
                collection_name, pk_name,
            )
        if embedding_dim is not None and embedding_dim != expected_dim:
            logger.warning(
                "⚠️  Milvus 集合 %s embedding 维度不一致 (expected=%d, actual=%d)，"
                "请手动 drop 旧集合后重新 ingest 全量数据",
                collection_name, expected_dim, embedding_dim,
            )

    # ─────────────────────────────── 生命周期 ────────────────────────────────

    def close(self):
        if self._pg:
            try:
                self._pg.close()
            except Exception as e:
                logger.warning("⚠️  PG 关闭失败: %s", e)
        if self._kafka_producer:
            try:
                self._kafka_producer.flush()
                self._kafka_producer.close()
            except Exception as e:
                logger.warning("⚠️  Kafka 关闭失败: %s", e)
        if self._es:
            try:
                self._es.close()
            except Exception as e:
                logger.warning("⚠️  ES 关闭失败: %s", e)
        if self._milvus:
            try:
                self._milvus.close()
            except Exception as e:
                logger.warning("⚠️  Milvus 关闭失败: %s", e)
